### ISSUE TITLE
servicecatalog-puppet generate-shares

### ISSUE THREAD

**Issue by nikolaigauss** [awslabs/aws-service-catalog-puppet#289]
Please include a link to your expanded manifest, the full contents of your AWS CodeBuild output (see https://aws-service-catalog-puppet.readthedocs.io/en/latest/puppet/using_the_cli.html#export-puppet-pipeline-logs)

We're having problems with the generate-shares in one of our environments, attached you'll find 2 expanded-manifests

[Archive.zip](https://github.com/awslabs/aws-service-catalog-puppet/files/4886469/Archive.zip)

these manifest are exactly the same in terms of launches but the accounts differ, the problem we're having with the bad manifest is that every time we generate shares we got this error:

```
/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/luigi/parameter.py:286: UserWarning: Parameter "single_account" with value "None" is not of type string.
  warnings.warn('Parameter "{}" with value "{}" is not of type string.'.format(param_name, param_value))
Traceback (most recent call last):
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/bin/servicecatalog-puppet", line 8, in <module>
    sys.exit(cli())
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/cli.py", line 21, in generate_shares
    core.generate_shares(f)
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/core.py", line 111, in generate_shares
    sub_tasks = spoke_local_portfolios_task.generate_tasks(t.get('task_defs'))
  File "/Users/nicolas_piechocki/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/workflow/provisioning.py", line 880, in generate_tasks
    raise Exception("The configuration for this share does not include any target accounts")
Exception: The configuration for this share does not include any target accounts
```

Whereas in the good manifest, the generate shares task run successfully .

Validation is not working for both with the same error:

```
(venv) nikolaigauss@C02XC196JGH6:~/workspace/tshooting-puppet $ servicecatalog-puppet validate manifest-expanded-good.yaml
Traceback (most recent call last):
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/bin/servicecatalog-puppet", line 8, in <module>
    sys.exit(cli())
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/cli.py", line 159, in validate
    core.validate(f)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/core.py", line 637, in validate
    c.validate(raise_exception=True)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 155, in validate
    self._start_validate(self.source)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 202, in _start_validate
    self._validate(value, root_rule, path, done)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 236, in _validate
    self._validate_mapping(value, rule, path, done=None)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 496, in _validate_mapping
    self._handle_func(value, rule, path, done)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 259, in _handle_func
    ret = method(value, rule, path)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/puppet_schema_extensions.py", line 10, in ext_schema
    ensure_deploy_to_tags_exist(value)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/puppet_schema_extensions.py", line 28, in ensure_deploy_to_tags_exist
    f"{collection_type}.{collection_name} uses tag {tag_to_check} in deploy_to that doesn't exist"
AssertionError: launches.ads-iam-bastion-roles uses tag type:ads-bastion in deploy_to that doesn't exist
(venv) nikolaigauss@C02XC196JGH6:~/workspace/tshooting-puppet $ servicecatalog-puppet validate manifest-expanded-bad.yaml
Traceback (most recent call last):
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/bin/servicecatalog-puppet", line 8, in <module>
    sys.exit(cli())
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/cli.py", line 159, in validate
    core.validate(f)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/core.py", line 637, in validate
    c.validate(raise_exception=True)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 155, in validate
    self._start_validate(self.source)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 202, in _start_validate
    self._validate(value, root_rule, path, done)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 236, in _validate
    self._validate_mapping(value, rule, path, done=None)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 496, in _validate_mapping
    self._handle_func(value, rule, path, done)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/pykwalify/core.py", line 259, in _handle_func
    ret = method(value, rule, path)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/puppet_schema_extensions.py", line 10, in ext_schema
    ensure_deploy_to_tags_exist(value)
  File "/Users/nikolaigauss/git/aws/ServiceCatalogTools/ServiceCatalogPuppet/venv/lib/python3.7/site-packages/servicecatalog_puppet/puppet_schema_extensions.py", line 28, in ensure_deploy_to_tags_exist
    f"{collection_type}.{collection_name} uses tag {tag_to_check} in deploy_to that doesn't exist"
AssertionError: launches.ads-iam-bastion-roles uses tag type:ads-bastion in deploy_to that doesn't exist
```

# Steps to reproduce
1. Expand manifest
2. Generate shares via CLI

# Expected results
Pipeline to run successfully.

# Actual results
Pipeline won't pass generate-shares step.

**Comment by eamonnfaherty** [comment]
In your launches you have defined a launch named ads-iam-bastion-roles.  The framework is trying to launch this into accounts or OUs that are tagged ads-bastion but in your manifest there are none tagged that way.  This is erroneous behaviour and the fact that the 'good' one runs without error is actually the bug.  Could you please confirm you are using the same version of the framework with both sets of tests using the servicecatalog-puppet version command please.

However, this was a change made in a recent version and I have had feedback that this is breaking existing installs for customers so I have reverted it.  I am planning a release tomorrow including this change. The change also includes an already released change to the logging to reduce the noise substantially and a new makefile to make developing/testing/experimenting with the framework easier.

**Comment by nikolaigauss** [comment]
Hello!
I can confirm that both environments are using service catalog puppet version: 0.73.0. Do please let me know if further information is needed.

**Comment by eamonnfaherty** [comment]
this has now been resolved in 0.76.4.
